staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,633 @@
+// 1) Start with shuffle sharding, to pick a queue.
+// 2) Reject old requests that have been waiting too long
+// 3) Reject current request if there is not enough concurrency shares and
+// we are at max queue length
+// 4) If not rejected, create a packet and enqueue
+// returns true on a successful enqueue
+// returns false in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+func (qs *QueueSetImpl) TimeoutOldRequestsAndRejectOrEnqueue(hashValue uint64, handSize int32) *Packet {
+	// TODO(aaron-prindle) removing locking now and doing it all in Wait()
+	// qs.lock.Lock()
+	// defer qs.lock.Unlock()
+
+	//	Start with the shuffle sharding, to pick a queue.
+	queueIdx := qs.ChooseQueueIdx(hashValue, int(handSize))
+	queue := qs.queues[queueIdx]
+	// The next step is the logic to reject requests that have been waiting too long
+	qs.removeTimedOutPacketsFromQueue(queue)
+	// NOTE: currently timeout is only checked for each new request.  This means that there can be
+	// requests that are in the queue longer than the timeout if there are no new requests
+	// We think this is a fine tradeoff
+
+	// Create a packet and enqueue
+	pkt := &Packet{
+		DequeueChannel: make(chan bool, 1),
+		EnqueueTime:    qs.clk.Now(),
+		Queue:          queue,
+	}
+	if ok := qs.rejectOrEnqueue(pkt); !ok {
+		return nil
+	}
+	return pkt
+
+}
+
+// removeTimedOutPacketsFromQueue rejects old requests that have been enqueued
+// past the requestWaitLimit
+func (qs *QueueSetImpl) removeTimedOutPacketsFromQueue(queue *Queue) {
+	timeoutIdx := -1
+	now := qs.clk.Now()
+	pkts := queue.Packets
+	// pkts are sorted oldest -> newest
+	// can short circuit loop (break) if oldest packets are not timing out
+	// as newer packets also will not have timed out
+
+	// now - requestWaitLimit = waitLimit
+	waitLimit := now.Add(-qs.requestWaitLimit)
+	for i, pkt := range pkts {
+		if waitLimit.After(pkt.EnqueueTime) {
+			if qs.wg != nil {
+				qs.wg.Add(1)
+			}
+			pkt.DequeueChannel <- false
+			close(pkt.DequeueChannel)
+			// // TODO(aaron-prindle) verify this makes sense here
+			// get idx for timed out packets
+			timeoutIdx = i
+
+		} else {
+			break
+		}
+	}
+	// remove timed out packets from queue
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout pkt
+		removeIdx := timeoutIdx + 1
+
+		// remove all the timeout packets
+		queue.Packets = pkts[removeIdx:]
+		qs.DecrementPackets(removeIdx)
+	}
+}
+
+// DecrementPackets decreases the # of packets for the QueueSetImpl w/ lock
+func (qs *QueueSetImpl) DecrementPackets(i int) {
+	qs.numPacketsEnqueued -= i
+}
+
+// GetRequestsExecuting gets the # of requests which are "executing":
+// this is the# of requests/packets which have been dequeued but have not had
+// finished (via the FinishPacket method invoked after service)
+func (qs *QueueSetImpl) GetRequestsExecuting() int {
+	total := 0
+	for _, queue := range qs.queues {
```

Comment:
This could cost O(1) instead of O(numQueues) if we incrementally maintain this quantity in `qs`.